graviton/abi_strategy.py

```diff
@@ -183,3 +204,161 @@ def get(self) -> PyTypes:
             1 << (cast(abi.UintType, self.abi_type).bit_size // 2)
         )
+
+
+class ABIArgsMod(Enum):
+    selector_byte_insert = auto()
+    selector_byte_delete = auto()
+    selector_byte_replace = auto()
+    parameter_delete = auto()
+    parameter_append = auto()
+
+
+class ABICallStrategy:
+    """
+    TODO: refactor to comport with ABIStrategy + Hypothesis
+    TODO: make this generic on the strategy type
+    """
+
+    append_args_type: abi.ABIType = abi.ByteType()
+
+    def __init__(
+        self,
+        contract: str,
+        argument_strategy: Type[ABIStrategy] = RandomABIStrategy,
+        *,
+        num_dryruns: int = 1,
+        handle_selector: bool = True,
+        abi_args_mod: Optional[ABIArgsMod] = None,
+    ):
+        """
+        contract - ABI Contract JSON
+
+        argument_strategy (default=RandomABIStrategy) - ABI strategy for generating arguments
+
+        num_dry_runs (default=1) - the number of dry runs to run
+            (generates different inputs each time)
+
+        handle_selector (default=True) - usually we'll want to let
+            `ABIContractExecutor.run_sequence()`
+            handle adding the method selector so this param.
+            But if set False: when providing `inputs`
+            ensure that the 0'th argument for method calls is the selector.
+            And when set True: when NOT providing `inputs`, the selector arg
+            at index 0 will be added automatically.
+
+        abi_args_mod (optional) - when desiring to mutate the args, provide an ABIArgsMod value
+        """
+        self.contract: abi.Contract = abi.Contract.from_json(contract)
+        self.argument_strategy: Optional[Type[ABIStrategy]] = argument_strategy
+        self.num_dryruns = num_dryruns
+        self.handle_selector = handle_selector
+        self.abi_args_mod = abi_args_mod
+
+    def abi_method(self, method: Optional[str]) -> abi.Method:
+        assert method, "cannot get abi.Method for bare app call"
+
+        return self.contract.get_method_by_name(method)
+
+    def method_signature(self, method: Optional[str]) -> Optional[str]:
+        """Returns None, for a bare app call (method=None signals this)"""
+        if method is None:
+            return None
+
+        return self.abi_method(method).get_signature()
+
+    def method_selector(self, method: Optional[str]) -> bytes:
+        assert method, "cannot get method_selector for bare app call"
+
+        return self.abi_method(method).get_selector()
+
+    def argument_types(self, method: Optional[str]) -> List[abi.ABIType]:
+        """
+        Argument types (excluding selector)
+        """
+        if method is None:
+            return []
+
+        return [cast(abi.ABIType, arg.type) for arg in self.abi_method(method).args]
+
+    def num_args(self, method: Optional[str]) -> int:
+        return len(self.argument_types(method))
+
+    def generate(self, method: Optional[str]) -> List[Sequence[PyTypes]]:
```

Comment:
nit: I would suggest renaming to `generate_inputs` to align with previous implementation in `blackbox.py`, or renaming to `generate_inputs_for_method` to be specific.

Reply:
woah, I made my comment above _without_ reading this first. JINX